src/NHibernate/Engine/Query/QueryPlanCache.cs

```diff
@@ -196,20 +196,23 @@ private class HQLQueryPlanKey : IEquatable<HQLQueryPlanKey>, IDeserializationCal
 			// hashcode may vary among processes, they cannot be stored and have to be re-computed after deserialization
 			[NonSerialized]
 			private int? _hashCode;
+
+			private readonly System.Type queryType;
 			private readonly System.Type queryTypeDiscriminator;
 
 			public HQLQueryPlanKey(string query, bool shallow, IDictionary<string, IFilter> enabledFilters)
-				: this(typeof(object), query, shallow, enabledFilters)
+				: this(typeof(object), typeof(object), query, shallow, enabledFilters)
 			{
 			}
 
 			public HQLQueryPlanKey(IQueryExpression queryExpression, bool shallow, IDictionary<string, IFilter> enabledFilters)
-				: this(queryExpression.GetType(), queryExpression.Key, shallow, enabledFilters)
+				: this(queryExpression.Type, queryExpression.GetType(), queryExpression.Key, shallow, enabledFilters)
 			{
 			}
 
-			protected HQLQueryPlanKey(System.Type queryTypeDiscriminator, string query, bool shallow, IDictionary<string, IFilter> enabledFilters)
+			protected HQLQueryPlanKey(System.Type queryType, System.Type queryTypeDiscriminator, string query, bool shallow, IDictionary<string, IFilter> enabledFilters)
```

Comment:
I have attempted to reproduce the issue, to no avail. The `query` parameter was always enough to distinguish the cases, without the need of this additional `queryType` parameter.
I have mapped two entities having the same type name and build in different dynamic assemblies, and I have build LINQ queries on them: no troubles. The `query` parameter contains the entity name instead of the type name, and so they are distinguished. I had to map them with different entity names, otherwise the mapping would not compile.

I have put this test case in a PR on your branch: Ablu/nhibernate-core#1

Reply:
I have done some checks in the history, entity names were not in the `query` parameter before [NH-2218](https://nhibernate.jira.com/browse/NH-2218). Was this issue reproduced with a version earlier than NH v4.1?

Reply:
The reproduction should not even require any special mappings. We hit the issues with an anonymous type (so doing `from collection select new { x }`), which was generated by the Roslyn scripting API. The API generates new assemblies for each evaluation. I did not test it with the very latest NH, but the bug should still occur given the code. I unfortunately have no time left this week, but I will try to take a look at it next week!

Reply:
Then your issue looks heavily like a duplicate of #1526, fixed by #1532 in v5.1.